### PR TITLE
Implement node Id interface

### PR DESCRIPTION
**Pull request by cache-nez**

### CODE REVIEW

File: id/mod.go [hunk 1]
```diff
@@ -1,7 +1,60 @@
...
 
+import (
+	"crypto/sha256"
+	"go.dedis.ch/dela/mino"
+	"math/big"
+)
+
 type Id interface {
-	GetLength() byte
+	GetLength() int
```
Comment (gnarula) [inline id/mod.go]: Consider removing the `Get` prefix here. It's idiomatic in Go to have a getter without the `Get` prefix in method names.

File: id/mod.go [hunk 2]
```diff
@@ -1,7 +1,60 @@
...
 	GetBase() byte
 	GetDigit(pos int) byte
-}
\ No newline at end of file
+}
+
+type ArrayId struct {
```
Comment (gnarula) [inline id/mod.go]: `ArrayId` => `ArrayNodeID`

File: id/mod.go [hunk 3]
```diff
@@ -1,7 +1,60 @@
 package id
 
+import (
+	"crypto/sha256"
+	"go.dedis.ch/dela/mino"
+	"math/big"
+)
+
 type Id interface {
```
Comment (gnarula) [inline id/mod.go]: `Id` to `NodeID`

File: id/mod.go [hunk 4]
```diff
@@ -1,7 +1,60 @@
...
-	GetLength() byte
+	GetLength() int
 	GetBase() byte
 	GetDigit(pos int) byte
-}
\ No newline at end of file
+}
+
+type ArrayId struct {
+	Id   []byte
```
Comment (gnarula) [inline id/mod.go]: nit, `ID` here

File: id/mod.go [hunk 5]
```diff
@@ -1,7 +1,60 @@
...
+
+func (id ArrayId) GetBase() byte {
+	return id.Base
+}
+
+func (id ArrayId) GetDigit(pos int) byte {
+	return id.Id[pos]
+}
+
+func hash(addr mino.Address) (h *big.Int) {
```
Comment (gnarula) [inline id/mod.go]: Replace this with a non named return please, i.e. have a `return h` at the end

File: id/mod.go [hunk 6]
```diff
@@ -1,7 +1,60 @@
...
+	curDigit := big.NewInt(0)
+	id := make([]byte, len)
+	for i := 0; i < len; i++ {
+		id[i] = byte(curDigit.Mod(h, bigBase).Int64())
+		h.Div(h, bigBase)
+	}
+	return ArrayId{id, base}
+}
+
+func (id ArrayId) GetLength() int {
```
Comment (gnarula) [inline id/mod.go]: Please add comments for all public methods

File: id/mod.go [hunk 7]
```diff
@@ -1,7 +1,60 @@
...
\ No newline at end of file
+}
+
+type ArrayId struct {
+	Id   []byte
+	Base byte
+}
+
+// Constructs an Id by taking a hash of the address modulo (base ^ len), then presenting the id as an array of digits.
+func MakeArrayId(address mino.Address, base byte, len int) Id {
```
Comment (gnarula) [inline id/mod.go]: A more idomatic name would be `NewArrayNodeID` with a return type of `ArrayNodeID`. Please ensure the comment wraps around 80 chars.

Would be useful to have a unit test for this

File: id/prefix.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package id
+
+type Prefix interface {
+	GetLength() int
```
Comment (gnarula) [inline id/prefix.go]: Nit, consider removing `Get`

File: id/prefix.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	isPrefixOf(Id) bool
```
Comment (gnarula) [inline id/prefix.go]: Is there a reason this method is private?

File: id/prefix.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	Append(byte) Prefix
+}
+
+// Digits of the prefix have to be stored in a string, because this structure is used as a key in the routing table map.
+type PrefixImpl struct {
```
Comment (gnarula) [inline id/prefix.go]: Perhaps `StringPrefix` might be a better name

**Review by gnarula — CHANGES_REQUESTED** [review]
Added some comments about Go conventions in general. Please consider having a unit test for `MakeArrayId`. Looks good otherwise.

File: id/mod.go (hunk 1 above)
Comment (cache-nez) [reply id/mod.go]: Done, thank you! Does this also apply to `GetDigit`, which is not a simple getter?

File: id/prefix.go (hunk 9 above)
Comment (cache-nez) [reply id/prefix.go]: No, thank you for the catch!

File: id/mod.go (hunk 4 above)
Comment (cache-nez) [reply id/mod.go]: Left it as a private member for now. Do you think it makes sense to make it public?

**Comment by gnarula** [comment]
Thanks for adding the tests.One small comment following Noemien's remark in the other PR. Comments in Go begin with the method name. Ready to be merged after that :)

**Review by gnarula — APPROVED** [review]

**Pull request merged**